decoder.cpp

```diff
@@ -190,6 +198,108 @@ bool Decoder::readImageResources()
   return (length == 0);
 }
 
+double Decoder::getTypeNumber(const std::unique_ptr<OSType>& v)
+{
+  if (v->type() == OSTypeKey::LargeInteger) {
+    return static_cast<const OSTypeLargeInt*>(v.get())->value;
+  }
+  else if (v->type() == OSTypeKey::Long) {
+    return static_cast<const OSTypeInt*>(v.get())->value;
+  }
+  else if (v->type() == OSTypeKey::Double) {
+    return static_cast<const OSTypeDouble*>(v.get())->value;
+  }
+  throw std::runtime_error("wrong descriptor type");
+};
+
+// the parsing used here is written thanks to the description given here
+// https://community.adobe.com/t5/photoshop-ecosystem-discussions/how-to-get-animation-frame-id/m-p/8790051#M68134
+std::unique_ptr<OSTypeDescriptor> Decoder::readAnimatedDataSection()
+{
+  read32(); // data length
+
+  const uint32_t descVersion = read32();
+  if (descVersion != 16)
+    return nullptr;
+
+  auto desc = parseDescriptor();
+  if (!desc)
+    return nullptr;
+
+  using DescriptorMap = OSTypeDescriptor::DescriptorMap;
+  const DescriptorMap& descrs = desc->descriptors;
+  const DescriptorMap::const_iterator frameStateIter =
+    descrs.find("FSts");
+  const DescriptorMap::const_iterator framesIter =
+    descrs.find("FrIn");
+
+  uint32_t activeFrameIndex = 0;
+  if (frameStateIter != descrs.cend() &&
+    frameStateIter->second->type() == OSTypeKey::List) {
```

Comment:
There are too many cases where an item from the descriptors map is queried. We have some options:

1) Creating a DescriptorMap class (a wrapper for the `std::map`) with some extra options to make it easier to get a value, e.g.:
```c++
class DescriptorMap {
public:
  ...easier to use interface including something like:
  template<typename T>
  const T& getValue(key);
private:
  std::map<std::string, std::unique_ptr<OSType>> m_items;
};
```
Another thing I think is that the key can be an uint32 (not a std::string), because keys are a 32-bit value.

Reply:
> Another thing I think is that the key can be an uint32 (not a std::string), because keys are a 32-bit value.

According to the [descriptor spec](https://www.adobe.com/devnet-apps/photoshop/fileformatashtml/#50577411_21585), the `Key` is in two forms, an integer or a string. [Integers can be converted to strings](https://github.com/aseprite/psd/blob/3a6dbedf76f1c86b5540bcc34c815c32e0031931/image_resources.cpp#L118) but not the other way round, since some of these strings are basically strings (I'll need to double check this part to be sure) and do not have integer equivalent.

I'll see to the first bit of your comment and make necessary changes.

Reply:
Oh sorry, you are right, I was confused by the OSType key 👍